internal/ethapi/api.go

```diff
@@ -916,16 +921,32 @@ func DoEstimateGas(ctx context.Context, b Backend, args CallArgs, blockNrOrHash
 		}
 		hi = block.GasLimit()
 	}
+	// Recap the highest gas limit with account's avaliable balance.
+	if args.GasPrice != nil && args.GasPrice.ToInt().Uint64() != 0 {
+		state, _, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+		if err != nil {
+			return 0, err
+		}
+		balance := new(big.Int)
+		balance.Set(state.GetBalance(*args.From)) // from can't be nil
+		if args.Value != nil {
+			if args.Value.ToInt().Cmp(balance) >= 0 {
+				return 0, errors.New("insufficient funds for transfer")
+			}
+			balance.Sub(balance, args.Value.ToInt())
+		}
+		allowance := new(big.Int).Div(balance, args.GasPrice.ToInt())
+		if hi > allowance.Uint64() {
+			hi = allowance.Uint64()
```

Comment:
We should also add a log here:

```go
log.Warn("Gas estimation capped by limited funds", "original", hi, "fundable", allowance)
```

This way if someone hits this then they at least see some meaningful warning too, not just an estimation failure.

Reply:
Maybe also add it to the simulated backend, just for completion's sake